examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py

```diff
@@ -0,0 +1,249 @@
+    # and additional transformations to perform to each asset spec (e.g. adding metadata).
+    serializable_data = serializable_data_from_defs(
+        definitions=defs,
+        migration_state=migration_state,
+        airflow_instance=instance,
+    )
+
+    serializable_specs = []
+    # Construct serializable specs for each dag.
+    for dag in dag_infos.values():
+        spec = get_dag_asset_spec(
+            airflow_instance=instance,
+            task_asset_keys_in_dag=serializable_data.keys_in_dag.get(dag.dag_id, set()),
+            downstreams_asset_dependency_graph=serializable_data.downstreams_asset_graph,
+            dag_info=dag,
+        )
+        serializable_specs.append(spec)
+        for dep in spec.deps:
+            serializable_data.downstreams_asset_graph[dep.asset_key].add(dag.dag_asset_key)
+
+    return _SerializedData(
+        spec_transformations=serializable_data.spec_transformations,
+        specs=serializable_specs,
+    )
+
+
+def definitions_from_metadata(
+    definitions_metadata: "_SerializedData",
+    defs: Definitions,
+    instance: AirflowInstance,
+    sensor_minimum_interval_seconds: int,
+) -> Definitions:
+    assets_defs = construct_all_assets(
+        definitions=defs,
+        serialized_data=definitions_metadata,
+    )
+    return defs_with_assets_and_sensor(
+        defs, assets_defs, instance, sensor_minimum_interval_seconds, definitions_metadata
+    )
+
+
+def defs_with_assets_and_sensor(
+    defs: Definitions,
+    assets_defs: List[AssetsDefinition],
+    instance: AirflowInstance,
+    sensor_minimum_interval_seconds: int,
+    definitions_metadata: "_SerializedData",
+) -> Definitions:
+    airflow_sensor = build_airflow_polling_sensor(
+        airflow_instance=instance, minimum_interval_seconds=sensor_minimum_interval_seconds
+    )
+    return Definitions(
+        assets=assets_defs,
+        asset_checks=defs.asset_checks if defs else None,
+        sensors=[airflow_sensor, *defs.sensors] if defs and defs.sensors else [airflow_sensor],
+        schedules=defs.schedules if defs else None,
+        jobs=defs.jobs if defs else None,
+        executor=defs.executor if defs else None,
+        loggers=defs.loggers if defs else None,
+        resources=defs.resources if defs else None,
+    ).with_reconstruction_metadata(
+        {
+            f"{AIRFLOW_SOURCE_METADATA_KEY_PREFIX}/{instance.name}": serialize_value(
+                definitions_metadata
+            )
+        }
+    )
+
+
+@whitelist_for_serdes(
+    field_serializers={
+        "spec_transformations": TransformationDataSerializer,
+    }
+)
+@record
+class _SerializedData:
+    """Data that will be serialized and cached to avoid repeated calls to the Airflow API,
+    and repeated scans of passed-in Definitions objects.
+    """
+
+    spec_transformations: Dict[AssetKey, "AssetSpecTransformationData"]
+    specs: Sequence[SerializableAssetSpec]
```

Comment:
This is very confusing because it appears that you are unnecessarily building both things when you could apply the transformations to the specs right here.

I think the problem here is that `AssetSpecTransformationData` is too general of name. What this really is metadata and tags that you derive from information in the airflow instance.

PerAssetKeyAirflowMetadata
AssetInfoFromAiflow

or something like that would make this more clear how the mechanism here is that you are taking in passed `AssetSpec` and then applying metadata and tags only accessible from the airflow instance

Reply:
I think it's important to make the note that it's data specific to assets which already exist in the passed in definitions, vs assets which have yet to be created. But otherwise agree with everything you said.